connections/hole-punching.md

```diff
@@ -0,0 +1,288 @@
+the (TCP or UDP) network socket, e.g. controlling port-reuse. The libp2p
+implementation controls the full transport security stack. Furthermore, unless
+run behind a Firewall or NAT, _non-browser_ platform hosts can be directly
+dialed through their publicly reachable addresses.
+
+On the contrary, on _browser_ platforms a libp2p implementation does not manage the
+network socket nor does it control the transport security stack. In addition to
+these restrictions, hosts running on _browser_ platforms can not be directly
+dialed.
+
+### Implementations
+
+There are multiple libp2p implementations, e.g. go-libp2p, js-libp2p and
+rust-libp2p. It is important to keep in mind that in some cases platforms and
+implementations are not matched one-to-one, e.g. the two implementations
+js-libp2p and rust-libp2p can run on both the _browser_ and _non-browser_
+platform types.
+
+### Protocol types
+
+#### Transport
+
+A transport protocol allows basic host-to-host communication. Examples are TCP,
+Websockets building on top of TCP, as well as the protocols layered over UDP
+namely QUIC and WebRTC.
+
+#### Relay
+
+A relay protocol allows two clients to communicate, routing data via a relay
+server, instead of using a direct connection. Examples of relay protocols are
+[TURN], [circuit relay v1][circuit-relay-v1] and [circuit relay
+v2][circuit-relay-v2].
+
+#### STUN-like
+
+STUN-like protocols allow a host to detect whether they are running behind a
+firewall and/or NAT and if so they enable a host to discover their perceived
+public IP addresses as well as their port mapping. Examples of STUN-like
+protocols are [STUN] (hence the name) and [AutoNAT] in combination with
+[Identify].
+
+#### Coordination
+
+Coordination protocols run on top of some relay protocol enabling two hosts to
+coordinate the hole punching process. Examples of coordination protocols are the
+[Session Description Protocol (SDP)][SDP] and [Direct Connection Upgrade through
+Relay][DCUTR].
+
+### Discovery mechanism
+
+A discovery mechanism allows a node to discover peers in the network. A node can
+use such mechanism e.g. to find peers at random, to learn external addresses of
+peers, or more targeted to discover peers that provide a specific service. Among
+others one can build a discovery mechanism on top of the libp2p [Kademlia] and
+[Gossipsub] protocol. [IPFS's usage of a Kademlia DHT][ipfs-kademlia] is a
+good example of a discovery mechanism.
+
+Further details on discovery mechanisms are out of scope for this document. The
+descriptions below will assume a generic _external discovery mechanism_ to be
+available.
+
+## Vision
+
+When it comes to establishing direct connections one needs to differentiate by
+platform type (browser / non-browser) as as well as routability (public /
+private). The below describes on a high level how to establish direct
+connections across the various permutations of the two dimensions.
```

Comment:
Perhaps we could replace `Non-Browser` with `Native`?

```suggestion
When it comes to establishing direct connections one needs to differentiate by
runtime type (JS in browser / native code) as as well as routability (public /
private). The below describes on a high level how to establish direct
connections across the various permutations of the two dimensions.
```

Reply:
Still agree with the standpoint that `Non-Browser` is not ideal, though I am not sure `Native` is a good replacement. To me `Native` is something _native_ to a specific OS, Architecture or Processor similar to the definition on Wikipedia:

> In computing, native software or data-formats are those that were designed to run on a particular operating system. In a more technical sense, native code is code written specifically for a certain processor.[1] In contrast, cross-platform software can be run on multiple operating systems and/or computer architectures. 

https://en.wikipedia.org/wiki/Native_(computing)

This would not hold for libp2p NodeJS given that OS or Processor specific properties are abstracted via the V8 virtual machine.

Does that reasoning make sense @lidel?